packages/vite/src/node/plugins/esbuild.ts

```diff
@@ -234,3 +199,195 @@ function prettifyMessage(m: Message, code: string): string {
 }
+
+// modified from <https://github.com/TypeStrong/tsconfig/blob/v7.0.0/src/tsconfig.ts#L75-L95>
+
+/**
+ * Copyright (c) 2015 TypeStrong
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+async function findTSConfig(dir: string): Promise<string | void> {
+  const configFile = path.resolve(dir, 'tsconfig.json')
+
+  return stat(configFile).then((stats) => {
+    if (isFile(stats)) {
+      return configFile
+    }
+
+    const parentDir = path.dirname(dir)
+
+    if (dir === parentDir) {
+      return
+    }
+
+    return findTSConfig(parentDir)
+  })
+}
+
+/**
+ * Check if a file exists.
+ */
+function stat(filename: string): Promise<fs.Stats | void> {
+  return new Promise<fs.Stats | void>((resolve, reject) => {
+    fs.stat(filename, (err, stats) => {
+      return err ? resolve(undefined) : resolve(stats)
+    })
+  })
+}
+
+/**
+ * Check filesystem stat is a directory.
+ */
+function isFile(stats: fs.Stats | void) {
+  return stats
+    ? (stats as fs.Stats).isFile() || (stats as fs.Stats).isFIFO()
+    : false
+}
+
+// from <https://github.com/TypeStrong/tsconfig/pull/31>
+// by @dominikg
+
+/**
+ * replace dangling commas from pseudo-json string with single space
+ *
+ * limitations:
+ * - pseudo-json must not contain comments, use strip-json-comments before
+ * - only a single dangling comma before } or ] is removed
+ *   stripDanglingComma('[1,2,]') === '[1,2 ]
+ *   stripDanglingComma('[1,2,,]') === '[1,2, ]
+ *
+ * implementation heavily inspired by strip-json-comments
+ */
+function stripDanglingComma(jsonString: string) {
+  /**
+   * Check if char at qoutePosition is escaped by an odd number of backslashes preceding it
+   */
+  function isEscaped(jsonString: string, quotePosition: number) {
+    let index = quotePosition - 1
+    let backslashCount = 0
+
+    while (jsonString[index] === '\\') {
+      index -= 1
+      backslashCount += 1
+    }
+
+    return Boolean(backslashCount % 2)
```

Comment:
**suggestion:** This is not good readable 😕 
We should use `backslashCount % 2 === 1` instead

```suggestion
    return backslashCount % 2 === 1
```